### PR TITLE
Tech debt

### LINKED ISSUE #1542: Can't update IMA due to gnosis safe api url change
```[tasklist]
### Tasks
- [ ] https://github.com/skalenetwork/IMA/issues/1638
```

### LINKED ISSUE #1638: Update dependencies related to gnosis-safe-sdk
Gnosis updated their sdk from `safe-core-sdk` to `protocol-kit`. In IMA upgrade script we still use old SDK, which causes problems with upgrade procedure via multisig. We should update dependencies to use `protocol-kit` instead of deprecated `safe-core-sdk`.

https://www.npmjs.com/package/@gnosis.pm/safe-core-sdk

### PR DESCRIPTION
**Pull request by yavrsky**
What was done:
- updated dependencies, particularly moved to ethers V6 and upgrade-tools V3
- fixed upgrade script for mainnet part for compatibility with latest version of upgrade-tools
- fixed deploy script
- fixed tests

What wasn't done:
- fixing deploy and upgrade scripts for mainnet, added a separate task as tech debt 
https://github.com/skalenetwork/IMA/issues/1720

### CODE REVIEW

File: gas/calculateGas.ts [hunk 1]
```diff
@@ -155,17 +155,17 @@ describe("Gas calculation", () => {
         schainsInternal = await (await ethers.getContractFactory("SchainsInternal")).deploy() as SchainsInternal;
         skaleVerifier = await (await ethers.getContractFactory("SkaleVerifierMock")).deploy() as SkaleVerifierMock;
         wallets = await (await ethers.getContractFactory("Wallets")).deploy() as Wallets;
-        await contractManager.connect(deployer).setContractsAddress("KeyStorage", keyStorage.address);
-        await contractManager.connect(deployer).setContractsAddress("Nodes", nodes.address);
-        await contractManager.connect(deployer).setContractsAddress("Schains", schains.address);
-        await contractManager.connect(deployer).setContractsAddress("SchainsInternal", schainsInternal.address);
-        await contractManager.connect(deployer).setContractsAddress("SkaleVerifier", skaleVerifier.address);
-        await contractManager.connect(deployer).setContractsAddress("Wallets", wallets.address);
+        await contractManager.connect(deployer).setContractsAddress("KeyStorage", await keyStorage.getAddress());
```
Comment (DimaStebaev) [inline gas/calculateGas.ts]: Just `keyStorage`. It's already `AddressLike`.

File: migrations/deployMainnet.ts [hunk 2]
```diff
@@ -43,205 +42,162 @@ export async function getManifestFile(): Promise<string> {
...
 async function getContractManager() {
     const defaultFilePath = "data/skaleManagerComponents.json";
-    const jsonData = JSON.parse(await fs.readFile(defaultFilePath)) as SkaleABIFile;
+    const jsonData = JSON.parse(await fs.readFile(defaultFilePath, "utf-8")) as SkaleABIFile;
     try {
-        const contractManagerAddress = jsonData.contract_manager_address as string;
-        const contractManagerABI = jsonData.contract_manager_abi as [];
-        return { address: contractManagerAddress, abi: contractManagerABI };
+        const { contract_manager_address, contract_manager_abi } = jsonData;
+        return { address: contract_manager_address as unknown as string, abi: contract_manager_abi as [] };
```
Comment (DimaStebaev) [inline migrations/deployMainnet.ts]: But it's not `unknown` does it?

File: migrations/deployMainnet.ts [hunk 3]
```diff
@@ -43,205 +42,162 @@ export async function getManifestFile(): Promise<string> {
...
-    "DepositBoxEth",
-    "DepositBoxERC20",
-    "DepositBoxERC721",
-    "DepositBoxERC1155",
-    "DepositBoxERC721WithMetadata"
-]
+    ...depositBoxes
+];
+
+async function deployContract(name: string, args: unknown[], initializer: string) {
```
Comment (DimaStebaev) [inline migrations/deployMainnet.ts]: Make `args` and `initializer` optional?

File: migrations/deploySchain.ts [hunk 4]
```diff
@@ -19,9 +19,12 @@
  */
 
 /**
- * @file deploySchain.js
+ * @file deploySchain.ts
  * @copyright SKALE Labs 2019-Present
  */
+// TODO: Remove this line after closing issue https://github.com/skalenetwork/IMA/issues/1720
```
Comment (DimaStebaev) [inline migrations/deploySchain.ts]: But this PR fixes 1720.

File: migrations/upgradeMainnet.ts [hunk 5]
```diff
@@ -38,14 +51,14 @@ class ImaMainnetUpgrader extends Upgrader {
...
+        this.transactions.push(Transaction.from({
+            to: await messageProxyForMainnet.getAddress(),
             data: messageProxyForMainnet.interface.encodeFunctionData("setVersion", [newVersion])
-        });
+        }));
     }
 
     // deployNewContracts = () => { };
-
+
```
Comment (DimaStebaev) [inline migrations/upgradeMainnet.ts]: oops

File: scripts/test_deploy.sh [hunk 6]
```diff
@@ -15,6 +15,8 @@ export PRIVATE_KEY_FOR_ETHEREUM=$PRIVATE_KEY
 export PRIVATE_KEY_FOR_SCHAIN=$PRIVATE_KEY
 
 yarn deploy-skale-manager-components
-yarn deploy-to-both-chains
+yarn deploy-to-mainnet
+# TODO: Uncomment after fixing deploy script for schain, related issue https://github.com/skalenetwork/IMA/issues/1720
```
Comment (DimaStebaev) [inline scripts/test_deploy.sh]: This PR fixes 1720.

File: scripts/test_upgrade.sh [hunk 7]
```diff
@@ -41,46 +43,51 @@ VERSION="$DEPLOYED_VERSION" \
...
+# end of TODO
 VERSION="$DEPLOYED_VERSION" npx hardhat run migrations/deployMainnet.ts --network localhost
 
-CHAIN_NAME_SCHAIN="Test" \
-VERSION="$DEPLOYED_VERSION" \
-URL_W3_S_CHAIN="$URL_W3_S_CHAIN" \
-PRIVATE_KEY_FOR_SCHAIN="$PRIVATE_KEY_FOR_SCHAIN" \
-npx hardhat run migrations/deploySchain.ts --network schain
+# TODO: uncomment upgrade schain test after closing issue https://github.com/skalenetwork/IMA/issues/1720
```
Comment (DimaStebaev) [inline scripts/test_upgrade.sh]: No way. Let's fix schain upgrade script.

File: test/CommunityLocker.ts [hunk 8]
```diff
@@ -80,58 +80,58 @@ describe("CommunityLocker", () => {
...
+        await messageProxyForSchain.postMessage(await communityLocker.getAddress(), mainnetHash, fakeCommunityPool, activateData);
         expect(await communityLocker.activeUsers(user.address)).to.be.equal(true);
-        await messageProxyForSchain.postMessage(communityLocker.address, mainnetHash, fakeCommunityPool, lockData);
+        await messageProxyForSchain.postMessage(await communityLocker.getAddress(), mainnetHash, fakeCommunityPool, lockData);
         expect(await communityLocker.activeUsers(user.address)).to.be.equal(false);
     });
 
     it("should set time limit per message", async () => {
-        expect(await communityLocker.timeLimitPerMessage(mainnetHash)).to.be.equal(300);
+        expect(await communityLocker.timeLimitPerMessage(mainnetHash)).to.be.equal(BigInt(300));
```
Comment (DimaStebaev) [inline test/CommunityLocker.ts]: `300n`

File: test/CommunityLocker.ts [hunk 9]
```diff
@@ -80,58 +80,58 @@ describe("CommunityLocker", () => {
...
         await communityLocker.setTimeLimitPerMessage("Mainnet", 0)
             .should.be.eventually.rejectedWith("Not enough permissions to set constant");
         await communityLocker.grantRole(await communityLocker.CONSTANT_SETTER_ROLE(), deployer.address);
         await communityLocker.setTimeLimitPerMessage("Mainnet", 0);
-        expect(await communityLocker.timeLimitPerMessage(mainnetHash)).to.be.equal(0);
+        expect(await communityLocker.timeLimitPerMessage(mainnetHash)).to.be.equal(BigInt(0));
```
Comment (DimaStebaev) [inline test/CommunityLocker.ts]: `0n`

File: tsconfig.json [hunk 10]
```diff
@@ -18,5 +18,19 @@
         "skipLibCheck": true
     },
     "include": ["**/*.ts"],
-    "exclude": ["node_modules", "example", "build", "test-tokens"]
+    "exclude": [
+        "node_modules",
+        "example",
+        "build",
+        "test-tokens",
+        // TODO: Remove these excludes after closing https://github.com/skalenetwork/IMA/issues/1720
```
Comment (DimaStebaev) [inline tsconfig.json]: Remove please

File: scripts/test_upgrade.sh [hunk 11]
```diff
@@ -57,30 +64,34 @@ npx hardhat run migrations/changeManifest.ts --network localhost
...
 cp .openzeppelin/unknown-*.json "$GITHUB_WORKSPACE/.openzeppelin"
 cp ./data/skaleManagerComponents.json "$GITHUB_WORKSPACE/data/"
+cp "data/proxyMainnet.json" "$GITHUB_WORKSPACE/data"
 cp "./data/ima-schain-$DEPLOYED_VERSION-manifest.json" "$GITHUB_WORKSPACE/data/"
-ABI_FILENAME_MAINNET="proxyMainnet.json"
-cp "data/$ABI_FILENAME_MAINNET" "$GITHUB_WORKSPACE/data"
 cp "data/$ABI_FILENAME_SCHAIN" "$GITHUB_WORKSPACE/data"
 cd "$GITHUB_WORKSPACE"
-rm -r --interactive=never "$DEPLOYED_DIR"
+rm -rf "$DEPLOYED_DIR"
```
Comment (DimaStebaev) [inline scripts/test_upgrade.sh]: Never use `rm -f` in scripts.

File: test/TokenManagerEth.ts [hunk 12]
```diff
@@ -45,15 +45,16 @@ import { deployEthErc20 } from "./utils/deploy/schain/ethErc20";
...
 
 import { expect } from "chai";
 import { deployMessageProxyForSchainTester } from "./utils/deploy/test/messageProxyForSchainTester";
 import { deployKeyStorageMock } from "./utils/deploy/test/keyStorageMock";
 
 const schainName = "TestSchain";
-const schainHash = stringKeccak256(schainName);
+const schainHash = ethers.id(schainName);
+
+/* eslint-disable @typescript-eslint/no-unused-expressions */
```
Comment (DimaStebaev) [inline test/TokenManagerEth.ts]: Why?

File: test/utils/helper.ts [hunk 13]
```diff
@@ -46,15 +46,18 @@ export function stringFromHex(value: string) {
     return str;
 }
 
-export function stringKeccak256(value: string): string {
-    return ethers.utils.solidityKeccak256(["string"], [value]);
+export function stringToHex(str: string) {
```
Comment (DimaStebaev) [inline test/utils/helper.ts]: Remove and use ethers instead.

File: scripts/test_upgrade.sh (hunk 7 above)
Comment (DimaStebaev) [reply scripts/test_upgrade.sh]: I resolve it but do it without any pleasure.

**Review by DmytroNazarenko — DISMISSED** [review]

**Review by DimaStebaev — DISMISSED** [review]

**Review by DmytroNazarenko — APPROVED** [review]

**Pull request merged**